metrics/network/cpu_statistics_iperf.sh

```diff
@@ -0,0 +1,111 @@
+SCRIPT_PATH=$(dirname "$(readlink -f "$0")")
+source "${SCRIPT_PATH}/../lib/common.bash"
+
+# Test name
+TEST_NAME="${TEST_NAME:-cpu information}"
+# Image name
+IMAGE="${IMAGE:-local-iperf}"
+# Dockerfile
+DOCKERFILE="${SCRIPT_PATH}/iperf_dockerfile/Dockerfile"
+# Client name
+CLIENT_NAME="${CLIENT_NAME:-client_iperf}"
+# Server name
+SERVER_NAME="${SERVER_NAME:-server_iperf}"
+# Target bandwidth
+BANDWIDTH="${BANDWIDTH:-1G}"
+# Time in seconds to transmit
+TIMEOUT="${TIMEOUT:-160}"
+# Sample period in seconds to measure
+SAMPLE_PERIOD="${SAMPLE_PERIOD:-120}"
+# File to save perf results
+TMP_FILE=$(mktemp cpuinfo.XXXXXXXXXX)
+# Time in seconds is required to let the benchmark settle down
+SETTLE_TIME="${SETTLE_TIME:-20}"
+
+function remove_tmp_file() {
+	rm -rf $TMP_FILE
+}
+
+function main() {
+	# Check tools/commands dependencies
+	cmds=("awk" "docker" "perf")
+
+	init_env
+	check_cmds "${cmds[@]}"
+	check_dockerfiles_images "$IMAGE" "$DOCKERFILE"
+
+	# Start iperf server configuration
+	# Set the TMPDIR to an existing tmpfs mount to avoid a 9p unlink error
+	local init_cmds="export TMPDIR=/dev/shm"
+	local server_cmd="$init_cmds && iperf -s"
+	docker run --runtime $RUNTIME --name $SERVER_NAME -d $IMAGE bash -c "$server_cmd"
+	local server_address=$(docker inspect --format "{{.NetworkSettings.IPAddress}}" $SERVER_NAME)
+
+	# Start iperf client
+	local client_cmd="$init_cmds && iperf -c $server_address -b $BANDWIDTH -t $TIMEOUT"
+	docker run --runtime $RUNTIME --name $CLIENT_NAME -d $IMAGE bash -c "$client_cmd"
+
+ 	# This time in seconds is required to let the benchmark settle down
+ 	sleep "$SETTLE_TIME"
+
+ 	# Retrieve qemu information
+ 	PIDS="$(pgrep -d ',' qemu)"
+
+ 	metrics_json_init
+
+ 	# Start collecting cpu information
+ 	sudo perf stat -a -o $TMP_FILE -e cycles -e instructions -p ${PIDS} sleep ${SAMPLE_PERIOD}
+
+ 	# Save configuration
+ 	metrics_json_start_array
+
+	local result_instructions="$(cat $TMP_FILE | grep -w instructions | awk '{print $1}' | tr ',' ' ')"
+	local instructions="$(for i in $(echo $result_instructions); do  [[ $max < $i ]] && max=$i && maxpos=$pos; ((pos++)); done; echo $max)"
+	local position_instruction="$(for i in $(echo $result_instructions); do  [[ $max < $i ]] && max=$i && maxpos=$pos; ((pos++)); done; echo $((maxpos + 1)))"
+
+	local cycles="$(cat $TMP_FILE | grep -w cycles | awk '{print $1}' | tr ',' ' ' | cut -d' ' -f$position_instruction)"
+
+	local instructions_per_cycle="$(cat $TMP_FILE | grep -w 'insn per cycle' | cut -d '#' -f2 | awk '{print $1}')"
+
+	local json="$(cat << EOF
+	{
+		"instructions per cycle": {
+			"Result" : $instructions_per_cycle,
+			"Units"  : "ins per cycle"
+		},
+		"cycles": {
+			"Result" : $cycles,
+			"Units"  : "cycles"
+		},
+		"instructions": {
+			"Result" : $instructions,
+			"Units"  : "instructions"
+		}
+	}
+EOF
+)"
+
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Results"
+	metrics_json_save
+	clean_env
+	trap remove_tmp_file EXIT
```

Comment:
This trap needs to be very very early - basically as soon as you've created the tmpfile, which probably means it has to be the first command in main(). pls read up on trap in the bash manual.

Reply:
@grahamwhaley, first I see that in your system is showing `insns per cycle` while mine is showing `ins per cycle` that is the reason why is empty there. I will change the `grep` for that.  Now as we discussed we need to select the highest `instruction` value with its correspondent `cycle` value. This is what is doing is locating the highest `instruction` and then match with it is `cycle`. Or what is the result that you want to see?

Reply:
argh, `ins` vs `insns` - I suspect that is going to be a perf version difference thing - you might just have to regexp the code to handle both of those.

For the instruction and cycle - those strings of comma separated numbers I believe are **not** strings of comma separated numbers - they are a whole single number (in the billions), printed in the American? comma separated format :-(  Sooo, what I expect to see is that number (such as from the above example - '378646182579' for the cycles.

We don't extract a max in this test. We don't currently run the test multiple times either - and if we did, I'd expect us to store all the individual results, and then let the CI system work out how it checked that (for instance, if it looked for the max or the mean etc.).

OK?